src/include/txn.h

```diff
@@ -88,11 +98,25 @@ struct __wt_txn_global {
 	 */
 	volatile uint64_t oldest_id;
 
-	/*
-	 * Prevents the oldest ID moving forwards while threads are scanning
-	 * the global transaction state.
-	 */
-	WT_RWLOCK scan_rwlock;
+#if TIMESTAMP_SIZE > 0
+	uint8_t commit_timestamp[TIMESTAMP_SIZE];
+	uint8_t read_timestamp[TIMESTAMP_SIZE];
+	uint8_t oldest_timestamp[TIMESTAMP_SIZE];
+	uint8_t pinned_timestamp[TIMESTAMP_SIZE];
+	bool has_oldest_ts, has_pinned_ts;
+#endif
+
+	WT_SPINLOCK id_lock;
+
+	/* Protects the leading edge of the active transaction window. */
+	WT_RWLOCK current_rwlock;
+
+#if 0
```

Comment:
No, I didn't get it working the other way.  I either need to make it work with two locks (my preference as long as they don't end up entangled), or go back to having just one.

Either way, I'm expecting some followup work to improve concurrency in the transaction subsystem.  In particular, I'm considering moving away from the model where we have a static array of `WT_TXN_STATE` structures that we have to scan regardless of whether the associated session is active, and instead have transactions hook themselves into one/more sorted lists when they start a transaction or pin a transaction ID or timestamp.

My plan is for that work to happen in a followup ticket, after these changes are merged.